pkg/verify/verify.go

```diff
@@ -0,0 +1,103 @@
+package verify
+
+import (
+	"crypto/sha256"
+	"crypto/sha512"
+	"encoding/hex"
+	"errors"
+	"fmt"
+	"hash"
+	"io"
+)
+
+var (
+	ErrNoHashes  = errors.New("verify: no hashes provided")
+	ErrShortData = errors.New("verify: data too short")
+)
+
+type ErrUnknownAlgorithm struct {
+	algorithm string
+}
+
+func (e *ErrUnknownAlgorithm) Error() string {
+	return fmt.Sprintf("verify: unknown hash algorithm %q", e.algorithm)
+}
+
+type ErrInvalidSize struct {
+	size int64
+}
+
+func (e *ErrInvalidSize) Error() string {
+	return fmt.Sprintf("verify: invalid data size: %d", e.size)
+}
+
+type ErrVerifyFailed struct {
+	hash   *Hash
+	actual string
+}
+
+func (e *ErrVerifyFailed) Error() string {
+	return fmt.Sprintf("verify: expected %s hash %q but got %q", e.hash.algorithm, e.hash.expected, e.actual)
+}
+
+type Hash struct {
+	hash      hash.Hash
+	algorithm string
+	expected  string
+}
+
+type Verifier struct {
+	hashes []*Hash
+	size   int64
+	reader *io.LimitedReader
+}
+
+func NewVerifier(hashes map[string]string, size int64) (*Verifier, error) {
+	if len(hashes) == 0 {
+		return nil, ErrNoHashes
+	}
+	if size <= 0 {
+		return nil, &ErrInvalidSize{size}
+	}
+	v := &Verifier{
+		hashes: make([]*Hash, 0, len(hashes)),
+		size:   size,
+	}
+	for algorithm, value := range hashes {
+		h := &Hash{algorithm: algorithm, expected: value}
+		switch algorithm {
+		case "sha256":
+			h.hash = sha256.New()
+		case "sha512":
+			h.hash = sha512.New()
+		case "sha512_256":
+			h.hash = sha512.New512_256()
+		default:
+			return nil, &ErrUnknownAlgorithm{algorithm}
```

Comment:
This error shouldn't exist, as it will prevent adding new hashes to manifests in the future.